src/cocotb/handle.py

```diff
@@ -878,15 +827,29 @@ def _set_value(
                 "Assigning list of length %d to object %s of length %d"
                 % (len(value), self._name, len(self))
             )
-        for val_idx, self_idx in enumerate(
-            self._range_iter(self._range[0], self._range[1])
-        ):
-            self[self_idx]._set_value(value[val_idx], action, schedule_write)
+        for elem, self_idx in zip(value, self.range):
+            self[self_idx]._set_value(elem, action, schedule_write)
+
+    def __getitem__(self, index: int) -> ChildObjectT:
+        if isinstance(index, slice):
+            raise IndexError("Slice indexing is not supported")
+        if index in self._sub_handles:
+            return self._sub_handles[index]
+        new_handle = self._handle.get_handle_by_index(index)
+        if not new_handle:
+            raise IndexError(f"{self._path} contains no object at index {index}")
+        path = self._path + "[" + str(index) + "]"
```

Comment:
```suggestion
        path = f"{self._path}[{index}]"
```
to match the other code using `key`.